### ISSUE TITLE
Add a function to manually send or resend (event trigger) email to the UI for Order and Customer

### ISSUE THREAD

**Issue by monrostar** [vendure-ecommerce/vendure#3016]
**Is your feature request related to a problem? Please describe.**
At the moment we need to add the ability to send some emails to our customers because sometimes they may specify the wrong email in their order and do not receive notification that the order needs to be paid or that the order has been paid. Due to the fact that we don't have a feature to manually send emails from admin panel, we needed a feature to allow our managers to send emails at any time just by clicking the button.

**Describe the solution you'd like**
I reviewed the available options and discovered an interesting approach.

We can extend `EmailEventHandler` and add a property to specify a UI block for displaying the list of emails to resend, along with a function to generate the event from the UI. For instance, we can introduce `setUiOptions` and add properties like `block(order, customer)` and a function to generate data for these events.

This way, you can automatically integrate events into pre-prepared UI blocks. When adding a new event, you only need to create an event generator function, which can then be incorporated into Vendure core.

This enhancement can be added to the default `EmailPlugin`.

```ts
interface EventUIHandlerOptions<T = Order | Customer> {
  entity: T
  handler: (ctx: RequestContext, injector: Injector, entity: T, languageCode?: LanguageCode) => EmailEventHandler
  filter: (ctx: RequestContext, injector: Injector, entity: T, languageCode?: LanguageCode) => boolean
}

// Add UI options to add a feature to manually send emails from the Vendure UI
private setUIOptions?: (handlerOptions: EventUIHandlerOptions) => void

get uiOptions(): EventUIHandlerOptions | undefined
```

**Describe alternatives you've considered**
From the alternatives I found only ways in which you need to manually add these blocks to the interface and process for each event request for sending separately

**Comment by monrostar** [comment]
@michaelbromley 

This is a feature I'm working on at the moment and I'd love to see it inside vendure core. Once my task is complete, I can create a PR to solve this problem for our company and also for the Vendure Community

**Comment by monrostar** [comment]
Only I have a problem. I can't figure out how to make the handler return type have `EventWithContext` instead of `EventWithAsyncData`

**Comment by michaelbromley** [comment]
Hi Eugene,

Thanks for this thoughtful suggestion. I agree that it would be good to support the ability to re-send emails. Here are my first thoughts based on the outline you provided. (note: I wrote this as I was reading and understanding the PR, so there is a bit of back-and-forth as I get a better idea of your concept).

### General implementation

I find the naming of the "handler" function unintuitive, and based on the `@return` annotation maybe this is a holdover from an earlier iteration? The purpose of that function is to return an Event right? That will then be picked up by the EmailPlugin and trigger a new email send.

~But there's nothing to stop you from returning an unrelated event right? That could then potentially trigger an entirely different EmailEventHandler. I know this would be a "strange" way to use that API but for sure people will start doing so.~ I see that the TS typings enforce that the same event type be returned 👍 

### Admin UI parts

Will the Admin UI need to have a coupling with the EmailPlugin now? Or do you plan to use a ui extension to implement the required UI parts?

### Danger of re-publishing events

~Another concern I would have would be unintended side-effects from re-publishing events "synthetically". For instance in your example we might have an unrelated plugin that performs some logic on that state transition. We don't necessarily want to perform that logic a second time just because we want to re-send an email.~

I just looked at the actual implementation and realized that you already thought of this and it is not an issue 👍 

### Naming

The `setUIOptions` method seems mis-named to me. As I understand it is dealing with the concept of re-sending, so maybe `resendOptions` could be a  better choice. 

**handler function**

As mentioned above I think this is mis-named. 

**filter function**

The part specifically dealing with the UI is the `filter`, which (per your doc block) decides whether to present the option for re-sending in the UI. If so then a name like `canResendFromUi` would be more explicit.

But also looking at the implementation in the PR, this function is _also_ used to prevent re-publishing the event. In which case maybe `canResend` is better as it applied to both UI and API usage.

**idea: do we need 2 separate functions?**

Thinking about these 2 functions further, I am wondering whether we really need 2 separate functions? What is we had a single function like this?

```ts
createEvent: (
        ctx: RequestContext,
        injector: Injector,
        entity: InstanceType<E>,
        languageCode?: LanguageCode,
    ) => Promise<InputEvent | false> | InputEvent | false;
```

then, if the given entity is not eligible for re-send for whatever reason, you return false. What are you thoughts on this approach?

### Summary

In summary I like the general direction of this proposal and think we should go ahead with adding it to the EmailPlugin.


> Also, I'd like to add an sms plugin. Absolutely the same as for EmailPlugin. I would like to have such a plugin inside Vendure core

> And also I want to add a function to translate text in a template by keys using i18n. With this, it will be possible to use just 1 template for different languages

Please open separate issues for these. Both sound interesting, I'd like to hear more detail on what you think this would involve.

**Comment by monrostar** [comment]
> Hi Eugene,
> 
> Thanks for this thoughtful suggestion. I agree that it would be good to support the ability to re-send emails. Here are my first thoughts based on the outline you provided. (note: I wrote this as I was reading and understanding the PR, so there is a bit of back-and-forth as I get a better idea of your concept).
> 
> ### General implementation
> I find the naming of the "handler" function unintuitive, and based on the `@return` annotation maybe this is a holdover from an earlier iteration? The purpose of that function is to return an Event right? That will then be picked up by the EmailPlugin and trigger a new email send.
> 
> ~But there's nothing to stop you from returning an unrelated event right? That could then potentially trigger an entirely different EmailEventHandler. I know this would be a "strange" way to use that API but for sure people will start doing so.~ I see that the TS typings enforce that the same event type be returned 👍
> 
> ### Admin UI parts
> Will the Admin UI need to have a coupling with the EmailPlugin now? Or do you plan to use a ui extension to implement the required UI parts?
> 
> ### Danger of re-publishing events
> ~Another concern I would have would be unintended side-effects from re-publishing events "synthetically". For instance in your example we might have an unrelated plugin that performs some logic on that state transition. We don't necessarily want to perform that logic a second time just because we want to re-send an email.~
> 
> I just looked at the actual implementation and realized that you already thought of this and it is not an issue 👍
> 
> ### Naming
> The `setUIOptions` method seems mis-named to me. As I understand it is dealing with the concept of re-sending, so maybe `resendOptions` could be a better choice.
> 
> **handler function**
> 
> As mentioned above I think this is mis-named.
> 
> **filter function**
> 
> The part specifically dealing with the UI is the `filter`, which (per your doc block) decides whether to present the option for re-sending in the UI. If so then a name like `canResendFromUi` would be more explicit.
> 
> But also looking at the implementation in the PR, this function is _also_ used to prevent re-publishing the event. In which case maybe `canResend` is better as it applied to both UI and API usage.
> 
> **idea: do we need 2 separate functions?**
> 
> Thinking about these 2 functions further, I am wondering whether we really need 2 separate functions? What is we had a single function like this?
> 
> ```ts
> createEvent: (
>         ctx: RequestContext,
>         injector: Injector,
>         entity: InstanceType<E>,
>         languageCode?: LanguageCode,
>     ) => Promise<InputEvent | false> | InputEvent | false;
> ```
> 
> then, if the given entity is not eligible for re-send for whatever reason, you return false. What are you thoughts on this approach?
> 
> ### Summary
> In summary I like the general direction of this proposal and think we should go ahead with adding it to the EmailPlugin.
> 
> > Also, I'd like to add an sms plugin. Absolutely the same as for EmailPlugin. I would like to have such a plugin inside Vendure core
> 
> > And also I want to add a function to translate text in a template by keys using i18n. With this, it will be possible to use just 1 template for different languages
> 
> Please open separate issues for these. Both sound interesting, I'd like to hear more detail on what you think this would involve.

Yeah, you got that right. At the moment I wrote the code I have just to test this idea, I tried to do it as soon as possible.

Now that you have approved this feature I will rename all methods and variables to be more in line with this function.

Today I'm gonna create a separate issues for SMS and translations

**Comment by monrostar** [comment]
> ### Admin UI parts
> Will the Admin UI need to have a coupling with the EmailPlugin now? Or do you plan to use a ui extension to implement the required UI parts?
I'm thinking of adding UI as custom component which will be inside EmailPlugin

**Comment by monrostar** [comment]
> **idea: do we need 2 separate functions?**
> 
> Thinking about these 2 functions further, I am wondering whether we really need 2 separate functions? What is we had a single function like this?
> 
> ```ts
> createEvent: (
>         ctx: RequestContext,
>         injector: Injector,
>         entity: InstanceType<E>,
>         languageCode?: LanguageCode,
>     ) => Promise<InputEvent | false> | InputEvent | false;
> ```
> 
> then, if the given entity is not eligible for re-send for whatever reason, you return false. What are you thoughts on this approach?

I think we need 2 separate functions to separate the validation function and the function to generate the Event because the CreateEvent function can take a bit longer to execute and I would like to reduce the API request time for the order-detail component. Well, it's better to separate them, it won't make things worse and it will be more convenient

**Comment by dlhck** [comment]
This feature request has been moved to GitHub Discussions for community voting.
Please upvote the discussion if you would like to see this implemented: https://github.com/vendurehq/vendure/discussions/4249
